File: backend/src/eams-parser.js
// 教务系统（EAMS）课表解析器。
// 算法思路参考开源项目「拾光课程表」（Apache-2.0），归属与致谢见仓库根目录 README。

const DEFAULT_TIME_SLOTS = [
  { number: 1, startTime: '07:50', endTime: '08:35' },
  { number: 2, startTime: '08:45', endTime: '09:30' },
  { number: 3, startTime: '09:50', endTime: '10:35' },
  { number: 4, startTime: '10:45', endTime: '11:30' },
  { number: 5, startTime: '11:31', endTime: '12:15' },
  { number: 6, startTime: '14:00', endTime: '14:45' },
  { number: 7, startTime: '14:55', endTime: '15:40' },
  { number: 8, startTime: '16:00', endTime: '16:45' },
  { number: 9, startTime: '16:55', endTime: '17:40' },
  { number: 10, startTime: '19:15', endTime: '20:00' },
  { number: 11, startTime: '20:10', endTime: '20:55' },
  { number: 12, startTime: '21:05', endTime: '21:50' }
]

function decodeEntities(value) {
  return String(value || '')
    .replace(/&quot;/g, '"')
    .replace(/&#39;|&apos;/g, "'")
    .replace(/&lt;/g, '<')
    .replace(/&gt;/g, '>')
    .replace(/&amp;/g, '&')
}

function splitArgs(argsString) {
  const args = []
  let current = ''
  let depth = 0
  let quote = null
  let escaped = false
  for (const character of argsString) {
    if (quote) {
      current += character
      if (escaped) {
        escaped = false
      } else if (character === '\\') {
        escaped = true
      } else if (character === quote) {
        quote = null
      }
    } else if (character === '"' || character === "'") {
      quote = character
      current += character
    } else if (character === '(' || character === '[' || character === '{') {
      depth += 1
      current += character
    } else if (character === ')' || character === ']' || character === '}') {
      depth -= 1
      current += character
    } else if (character === ',' && depth === 0) {
      args.push(current.trim())
      current = ''
    } else {
      current += character
    }
  }
  if (current.trim()) args.push(current.trim())
  return args
}

function stringLiteral(value) {
  if (!value || value.length < 2) return value || ''
  if (value[0] === '"' && value[value.length - 1] === '"') {
    try {
      return JSON.parse(value)
    } catch {
      return value.slice(1, -1)
    }
  }
  if (value[0] === "'" && value[value.length - 1] === "'") {
    return value.slice(1, -1).replace(/\\'/g, "'").replace(/\\\\/g, '\\')
  }
  return value
}

function resolveArg(argument, teacherNames, assistantName) {
  if (argument === undefined || argument === null || /^null$/i.test(argument)) return ''
  if (argument.includes('actTeacherName')) return teacherNames.join(',')
  if (argument.includes('actTeacherId')) return ''
  if (argument.includes('assistantName')) return assistantName
  return decodeEntities(stringLiteral(argument))
}

function resolveTaskActivityArgs(argsString, teacherNames, assistantName) {
  const args = splitArgs(argsString)
  if (args.length < 7) return null
  const rawName = resolveArg(args[3], teacherNames, assistantName)
  const courseName = rawName.replace(/\s*[（(][0-9A-Za-z.]{3,}[）)]\s*$/, '')
  const teacher = [resolveArg(args[1], teacherNames, assistantName), assistantName]
    .filter((value) => value && value.trim())
    .join(',')
  return {
    name: courseName,
    teacher,
    room: resolveArg(args[5], teacherNames, assistantName),
    validWeeks: resolveArg(args[6], teacherNames, assistantName)
  }
}

function parseTimetable(html) {
  if (typeof html !== 'string' || !html.trim()) throw new Error('课表源码为空')
  const unitCountMatch = html.match(/var\s+unitCount\s*=\s*(\d+)/)
  const marshalMatch = html.match(/marshalTable\((\d+)\s*,\s*(\d+)\s*,\s*(\d+)\)/)
  if (!unitCountMatch || !marshalMatch) {
    throw new Error('没有找到 EAMS TaskActivity 课表数据，请确认文件来自登录后的学生课表页')
  }
  const unitCount = Number(unitCountMatch[1])
  const from = Number(marshalMatch[1])
  const startWeek = Number(marshalMatch[2])
  const endWeek = Number(marshalMatch[3])
  if (!unitCount || startWeek < 1 || endWeek < startWeek) throw new Error('课表周次数据异常')

  const segments = html.split('new TaskActivity(')
  const rawCourses = []
  for (let index = 1; index < segments.length; index += 1) {
    const segment = segments[index]
    const callEnd = segment.indexOf(');')
    if (callEnd < 0) continue
    const rest = segment.slice(callEnd + 2)
    const indexPattern = /index\s*=\s*(\d+)\s*\*\s*unitCount\s*\+\s*(\d+)\s*;/g
    const indexMatches = [...rest.matchAll(indexPattern)]
    if (!indexMatches.length) continue

    const previous = segments[index - 1]
    const teacherDeclarations = [...previous.matchAll(/var\s+actTeachers\s*=\s*\[([^\]]*)\]/g)]
    const lastTeacherDeclaration = teacherDeclarations[teacherDeclarations.length - 1]
    const teacherNames = lastTeacherDeclaration
      ? [...lastTeacherDeclaration[1].matchAll(/name\s*:\s*"((?:\\.|[^"])*)"/g)]
          .map((match) => stringLiteral(`"${match[1]}"`))
      : []
    const assistantMatches = [...previous.matchAll(/var\s+assistantName\s*=\s*"((?:\\.|[^"])*)"/g)]
    const assistantName = assistantMatches.length
      ? stringLiteral(`"${assistantMatches[assistantMatches.length - 1][1]}"`)
      : ''
    const info = resolveTaskActivityArgs(segment.slice(0, callEnd), teacherNames, assistantName)
    if (!info || !/^[01]+$/.test(info.validWeeks)) continue

    const validStart = from + startWeek - 2
    let validWeeks = info.validWeeks
    if (validWeeks.slice(0, validStart).includes('1') && !validWeeks.slice(validStart).includes('1')) {
      validWeeks = `${validWeeks.slice(1)}0`
    }

    const weeks = []
    for (let week = startWeek; week <= endWeek; week += 1) {
      const validIndex = from + week - 2
      if (validIndex >= 0 && validIndex < validWeeks.length && validWeeks[validIndex] === '1') {
        weeks.push(week)
      }
    }
    if (!weeks.length) continue

    indexMatches.forEach((match) => {
      const day = Number(match[1]) + 1
      const section = Number(match[2]) + 1
      if (day >= 1 && day <= 7 && section >= 1 && section <= 12) {
        rawCourses.push({
          name: info.name,
          teacher: info.teacher,
          position: info.room,
          day,
          startSection: section,
          endSection: section,
          weeks: weeks.slice()
        })
      }
    })
  }
  const courses = mergeAndDistinctCourses(rawCourses)
  if (!courses.length) throw new Error('已识别课表结构，但没有解析到有效课程')
  return {
    format: 'zhusijiao-schedule',
    version: 1,
    school: '',
    name: '我的课表',
    semesterStart: '',
    totalWeeks: endWeek,
    timeSlots: DEFAULT_TIME_SLOTS.map((slot) => ({ ...slot })),
    courses
  }
}

function sameWeeks(first, second) {
  return first.join(',') === second.join(',')
}

function mergeAndDistinctCourses(courses) {
  if (!Array.isArray(courses) || !courses.length) return []
  const list = courses.map((course) => ({
    ...course,
    weeks: [...new Set((course.weeks || []).map(Number))].sort((a, b) => a - b)
  }))
  list.sort((a, b) => (
    a.name.localeCompare(b.name) ||
    a.teacher.localeCompare(b.teacher) ||
    a.position.localeCompare(b.position) ||
    a.day - b.day ||
    a.weeks.join(',').localeCompare(b.weeks.join(',')) ||
    a.startSection - b.startSection
  ))

  const consecutive = []
  let current = { ...list[0], weeks: [...list[0].weeks] }
  for (let index = 1; index < list.length; index += 1) {
    const next = list[index]
    const sameCourse = current.name === next.name && current.teacher === next.teacher &&
      current.position === next.position && current.day === next.day && sameWeeks(current.weeks, next.weeks)
    if (sameCourse && current.endSection + 1 === next.startSection) {
      current.endSection = next.endSection
    } else if (!(sameCourse && current.startSection === next.startSection && current.endSection === next.endSection)) {
      consecutive.push(current)
      current = { ...next, weeks: [...next.weeks] }
    }
  }
  consecutive.push(current)

  consecutive.sort((a, b) => (
    a.name.localeCompare(b.name) ||
    a.teacher.localeCompare(b.teacher) ||
    a.position.localeCompare(b.position) ||
    a.day - b.day ||
    a.startSection - b.startSection ||
    a.endSection - b.endSection
  ))
  const merged = []
  current = { ...consecutive[0], weeks: [...consecutive[0].weeks] }
  for (let index = 1; index < consecutive.length; index += 1) {
    const next = consecutive[index]
    const sameSlot = current.name === next.name && current.teacher === next.teacher &&
      current.position === next.position && current.day === next.day &&
      current.startSection === next.startSection && current.endSection === next.endSection
    if (sameSlot) {
      current.weeks = [...new Set([...current.weeks, ...next.weeks])].sort((a, b) => a - b)
    } else {
      merged.push(current)
      current = { ...next, weeks: [...next.weeks] }
    }
  }
  merged.push(current)
  return merged
}

function normalizeJsonPackage(value) {
  const source = value && value.schedule ? value.schedule : value
  if (!source || !Array.isArray(source.courses)) throw new Error('JSON 中没有 courses 数组')
  const courses = source.courses.map((course) => ({
    name: String(course.name || '').trim(),
    teacher: String(course.teacher || '').trim(),
    position: String(course.position || course.room || '').trim(),
    day: Number(course.day),
    startSection: Number(course.startSection),
    endSection: Number(course.endSection || course.startSection),
    weeks: Array.isArray(course.weeks) ? course.weeks.map(Number) : []
  })).filter((course) => (
    course.name && course.day >= 1 && course.day <= 7 &&
    course.startSection >= 1 && course.endSection <= 12 && course.weeks.length
  ))
  if (!courses.length) throw new Error('JSON 中没有有效课程')
  const maxWeek = Math.max(...courses.flatMap((course) => course.weeks))
  return {
    format: 'zhusijiao-schedule',
    version: 1,
    school: '',
    name: String(source.name || '我的课表').trim(),
    semesterStart: String(source.semesterStart || '').trim(),
    totalWeeks: Number(source.totalWeeks || maxWeek || 20),
    timeSlots: Array.isArray(source.timeSlots) && source.timeSlots.length
      ? source.timeSlots.slice(0, 12)
      : DEFAULT_TIME_SLOTS.map((slot) => ({ ...slot })),
    courses: mergeAndDistinctCourses(courses)
  }
}

function parseImportContent(content) {
  let source = String(content || '').trim()
  if (!source) throw new Error('导入内容为空')
  if (source.startsWith('```')) {
    source = source.replace(/^```(?:json|html|javascript)?\s*/i, '').replace(/\s*```$/, '')
  }
  if (source[0] === '{' || source[0] === '[') {
    try {
      return normalizeJsonPackage(JSON.parse(source))
    } catch (error) {
      if (error.message !== 'JSON 中没有 courses 数组' && error.message !== 'JSON 中没有有效课程') {
        throw new Error(`JSON 格式错误：${error.message}`)
      }
      throw error
    }
  }
  return parseTimetable(source)
}

module.exports = {
  DEFAULT_TIME_SLOTS,
  mergeAndDistinctCourses,
  parseImportContent,
  parseTimetable,
  splitArgs
}

File: app/src/main/java/com/zhusijiao/app/domain/EamsParser.kt
package com.zhusijiao.app.domain

import org.json.JSONObject
import org.json.JSONTokener
import kotlin.math.max
import kotlin.math.min

/**
 * 教务系统（EAMS）课表解析器。
 * 算法思路参考开源项目「拾光课程表」（Apache-2.0），归属与致谢见仓库根目录 README。支持两种输入：
 *  1) 含 `new TaskActivity(...)` 的课表页源码（HTML/JS）；
 *  2) 导出助手生成的 zhusijiao-schedule JSON。
 */
object EamsParser {

    class ParseException(message: String) : Exception(message)

    private data class TaskInfo(val name: String, val teacher: String, val room: String, val validWeeks: String)

    /** 可变工作项，便于两趟合并。 */
    private class Work(
        var name: String, var teacher: String, var position: String,
        var day: Int, var startSection: Int, var endSection: Int, var weeks: MutableList<Int>
    ) {
        fun toCourse() = Course(name, teacher, position, day, startSection, endSection, weeks.toList())
    }

    val DEFAULT_TIME_SLOTS: List<TimeSlot> = listOf(
        TimeSlot(1, "07:50", "08:35"),
        TimeSlot(2, "08:45", "09:30"),
        TimeSlot(3, "09:50", "10:35"),
        TimeSlot(4, "10:45", "11:30"),
        TimeSlot(5, "11:31", "12:15"),
        TimeSlot(6, "14:00", "14:45"),
        TimeSlot(7, "14:55", "15:40"),
        TimeSlot(8, "16:00", "16:45"),
        TimeSlot(9, "16:55", "17:40"),
        TimeSlot(10, "19:15", "20:00"),
        TimeSlot(11, "20:10", "20:55"),
        TimeSlot(12, "21:05", "21:50")
    )

    private val reUnitCount = Regex("var\\s+unitCount\\s*=\\s*(\\d+)")
    private val reMarshal = Regex("marshalTable\\((\\d+)\\s*,\\s*(\\d+)\\s*,\\s*(\\d+)\\)")
    private val reIndex = Regex("index\\s*=\\s*(\\d+)\\s*\\*\\s*unitCount\\s*\\+\\s*(\\d+)\\s*;")
    private val reActTeachers = Regex("var\\s+actTeachers\\s*=\\s*\\[([^\\]]*)\\]")
    private val reTeacherName = Regex("name\\s*:\\s*\"((?:\\\\.|[^\"])*)\"")
    private val reAssistant = Regex("var\\s+assistantName\\s*=\\s*\"((?:\\\\.|[^\"])*)\"")
    private val reBinary = Regex("^[01]+$")
    private val reCodeSuffix = Regex("\\s*[（(][0-9A-Za-z.]{3,}[）)]\\s*$")

    // ===== 字符串工具 =====

    /** JS String.slice 语义（支持越界与负索引钳制）。 */
    private fun String.jsSlice(start: Int, end: Int = length): String {
        val len = length
        var s = if (start < 0) max(len + start, 0) else min(start, len)
        var e = if (end < 0) max(len + end, 0) else min(end, len)
        if (e < s) e = s
        return substring(s, e)
    }

    private fun decodeEntities(value: String?): String = (value ?: "")
        .replace("&quot;", "\"")
        .replace("&#39;", "'").replace("&apos;", "'")
        .replace("&lt;", "<")
        .replace("&gt;", ">")
        .replace("&amp;", "&")

    /** 按顶层逗号切分实参，忽略引号与括号内的逗号。 */
    fun splitArgs(argsString: String): List<String> {
        val args = mutableListOf<String>()
        val current = StringBuilder()
        var depth = 0
        var quote: Char? = null
        var escaped = false
        for (ch in argsString) {
            if (quote != null) {
                current.append(ch)
                when {
                    escaped -> escaped = false
                    ch == '\\' -> escaped = true
                    ch == quote -> quote = null
                }
            } else if (ch == '"' || ch == '\'') {
                quote = ch
                current.append(ch)
            } else if (ch == '(' || ch == '[' || ch == '{') {
                depth += 1; current.append(ch)
            } else if (ch == ')' || ch == ']' || ch == '}') {
                depth -= 1; current.append(ch)
            } else if (ch == ',' && depth == 0) {
                args += current.toString().trim(); current.setLength(0)
            } else {
                current.append(ch)
            }
        }
        if (current.toString().trim().isNotEmpty()) args += current.toString().trim()
        return args
    }

    private fun stringLiteral(value: String): String {
        if (value.length < 2) return value
        if (value.startsWith("\"") && value.endsWith("\"")) {
            return try {
                val v = JSONTokener(value).nextValue()
                if (v is String) v else value.substring(1, value.length - 1)
            } catch (_: Exception) {
                value.substring(1, value.length - 1)
            }
        }
        if (value.startsWith("'") && value.endsWith("'")) {
            return value.substring(1, value.length - 1).replace("\\'", "'").replace("\\\\", "\\")
        }
        return value
    }

    private fun resolveArg(argument: String?, teacherNames: List<String>, assistantName: String): String {
        if (argument == null || argument.equals("null", ignoreCase = true)) return ""
        if (argument.contains("actTeacherName")) return teacherNames.joinToString(",")
        if (argument.contains("actTeacherId")) return ""
        if (argument.contains("assistantName")) return assistantName
        return decodeEntities(stringLiteral(argument))
    }

    private fun resolveTaskActivityArgs(argsString: String, teacherNames: List<String>, assistantName: String): TaskInfo? {
        val args = splitArgs(argsString)
        if (args.size < 7) return null
        val rawName = resolveArg(args[3], teacherNames, assistantName)
        val courseName = rawName.replace(reCodeSuffix, "")
        val teacher = listOf(resolveArg(args[1], teacherNames, assistantName), assistantName)
            .filter { it.isNotBlank() }
            .joinToString(",")
        return TaskInfo(
            name = courseName,
            teacher = teacher,
            room = resolveArg(args[5], teacherNames, assistantName),
            validWeeks = resolveArg(args[6], teacherNames, assistantName)
        )
    }

    // ===== 主解析 =====

    fun parseTimetable(html: String?): ParsedSchedule {
        if (html.isNullOrBlank()) throw ParseException("课表源码为空")
        val unitCountMatch = reUnitCount.find(html)
        val marshalMatch = reMarshal.find(html)
        if (unitCountMatch == null || marshalMatch == null) {
            throw ParseException("没有找到 EAMS TaskActivity 课表数据，请确认文件来自登录后的学生课表页")
        }
        val unitCount = unitCountMatch.groupValues[1].toIntOrNull() ?: 0
        val from = marshalMatch.groupValues[1].toIntOrNull() ?: 0
        val startWeek = marshalMatch.groupValues[2].toIntOrNull() ?: 0
        val endWeek = marshalMatch.groupValues[3].toIntOrNull() ?: 0
        if (unitCount == 0 || startWeek < 1 || endWeek < startWeek) throw ParseException("课表周次数据异常")

        val segments = html.split("new TaskActivity(")
        val rawCourses = mutableListOf<Course>()
        for (index in 1 until segments.size) {
            val segment = segments[index]
            val callEnd = segment.indexOf(");")
            if (callEnd < 0) continue
            val rest = segment.substring(callEnd + 2)
            val indexMatches = reIndex.findAll(rest).toList()
            if (indexMatches.isEmpty()) continue

            val previous = segments[index - 1]
            val teacherDeclarations = reActTeachers.findAll(previous).toList()
            val lastTeacher = teacherDeclarations.lastOrNull()
            val teacherNames = lastTeacher
                ?.let { reTeacherName.findAll(it.groupValues[1]).map { m -> stringLiteral("\"" + m.groupValues[1] + "\"") }.toList() }
                ?: emptyList()
            val assistantMatches = reAssistant.findAll(previous).toList()
            val assistantName = if (assistantMatches.isNotEmpty()) {
                stringLiteral("\"" + assistantMatches.last().groupValues[1] + "\"")
            } else ""

            val info = resolveTaskActivityArgs(segment.substring(0, callEnd), teacherNames, assistantName)
            if (info == null || !reBinary.matches(info.validWeeks)) continue

            val validStart = from + startWeek - 2
            var validWeeks = info.validWeeks
            if (validWeeks.jsSlice(0, validStart).contains('1') && !validWeeks.jsSlice(validStart).contains('1')) {
                validWeeks = validWeeks.jsSlice(1) + "0"
            }

            val weeks = mutableListOf<Int>()
            for (week in startWeek..endWeek) {
                val validIndex = from + week - 2
                if (validIndex in 0 until validWeeks.length && validWeeks[validIndex] == '1') weeks += week
            }
            if (weeks.isEmpty()) continue

            indexMatches.forEach { match ->
                val day = (match.groupValues[1].toIntOrNull() ?: 0) + 1
                val section = (match.groupValues[2].toIntOrNull() ?: 0) + 1
                if (day in 1..7 && section in 1..12) {
                    rawCourses += Course(info.name, info.teacher, info.room, day, section, section, weeks.toList())
                }
            }
        }
        val courses = mergeAndDistinctCourses(rawCourses)
        if (courses.isEmpty()) throw ParseException("已识别课表结构，但没有解析到有效课程")
        return ParsedSchedule(
            name = "我的课表",
            semesterStart = "",
            totalWeeks = endWeek,
            timeSlots = DEFAULT_TIME_SLOTS,
            courses = courses
        )
    }

    private fun sameWeeks(a: List<Int>, b: List<Int>) = a.joinToString(",") == b.joinToString(",")

    /** 先合并相邻节次，再对同课程同节次的周次去重合并。 */
    fun mergeAndDistinctCourses(courses: List<Course>): List<Course> {
        if (courses.isEmpty()) return emptyList()
        val list = courses.map {
            Work(it.name, it.teacher, it.position, it.day, it.startSection, it.endSection,
                it.weeks.distinct().sorted().toMutableList())
        }.sortedWith(
            compareBy({ it.name }, { it.teacher }, { it.position }, { it.day },
                { it.weeks.joinToString(",") }, { it.startSection })
        )

        // 第一趟：合并连续节次
        val consecutive = mutableListOf<Work>()
        var current = Work(list[0].name, list[0].teacher, list[0].position, list[0].day,
            list[0].startSection, list[0].endSection, list[0].weeks.toMutableList())
        for (index in 1 until list.size) {
            val next = list[index]
            val sameCourse = current.name == next.name && current.teacher == next.teacher &&
                current.position == next.position && current.day == next.day && sameWeeks(current.weeks, next.weeks)
            if (sameCourse && current.endSection + 1 == next.startSection) {
                current.endSection = next.endSection
            } else if (!(sameCourse && current.startSection == next.startSection && current.endSection == next.endSection)) {
                consecutive += current
                current = Work(next.name, next.teacher, next.position, next.day,
                    next.startSection, next.endSection, next.weeks.toMutableList())
            }
        }
        consecutive += current

        consecutive.sortWith(
            compareBy({ it.name }, { it.teacher }, { it.position }, { it.day },
                { it.startSection }, { it.endSection })
        )

        // 第二趟：合并同节次的周次
        val merged = mutableListOf<Work>()
        current = Work(consecutive[0].name, consecutive[0].teacher, consecutive[0].position, consecutive[0].day,
            consecutive[0].startSection, consecutive[0].endSection, consecutive[0].weeks.toMutableList())
        for (index in 1 until consecutive.size) {
            val next = consecutive[index]
            val sameSlot = current.name == next.name && current.teacher == next.teacher &&
                current.position == next.position && current.day == next.day &&
                current.startSection == next.startSection && current.endSection == next.endSection
            if (sameSlot) {
                current.weeks = (current.weeks + next.weeks).distinct().sorted().toMutableList()
            } else {
                merged += current
                current = Work(next.name, next.teacher, next.position, next.day,
                    next.startSection, next.endSection, next.weeks.toMutableList())
            }
        }
        merged += current
        return merged.map { it.toCourse() }
    }

    fun normalizeJsonPackage(value: JSONObject?): ParsedSchedule {
        val source = value?.optJSONObject("schedule") ?: value
            ?: throw ParseException("JSON 中没有 courses 数组")
        val coursesArray = source.optJSONArray("courses")
            ?: throw ParseException("JSON 中没有 courses 数组")
        val courses = (0 until coursesArray.length()).mapNotNull { i ->
            coursesArray.optJSONObject(i)
        }.map { o ->
            Course(
                name = o.optString("name").trim(),
                teacher = o.optString("teacher").trim(),
                position = o.optString("position").ifEmpty { o.optString("room") }.trim(),
                day = o.optInt("day"),
                startSection = o.optInt("startSection"),
                endSection = o.optInt("endSection", o.optInt("startSection")),
                weeks = o.optJSONArray("weeks").toIntList()
            )
        }.filter {
            it.name.isNotEmpty() && it.day in 1..7 && it.startSection >= 1 && it.endSection <= 12 && it.weeks.isNotEmpty()
        }
        if (courses.isEmpty()) throw ParseException("JSON 中没有有效课程")
        val maxWeek = courses.flatMap { it.weeks }.maxOrNull() ?: 0
        val slotsArray = source.optJSONArray("timeSlots")
        val timeSlots = if (slotsArray != null && slotsArray.length() > 0) {
            (0 until min(slotsArray.length(), 12)).map { TimeSlot.fromJson(slotsArray.optJSONObject(it) ?: JSONObject()) }
        } else DEFAULT_TIME_SLOTS
        return ParsedSchedule(
            name = source.optString("name", "我的课表").trim().ifEmpty { "我的课表" },
            semesterStart = source.optString("semesterStart").trim(),
            totalWeeks = source.optInt("totalWeeks").takeIf { it > 0 } ?: (maxWeek.takeIf { it > 0 } ?: 20),
            timeSlots = timeSlots,
            courses = mergeAndDistinctCourses(courses)
        )
    }

    fun parseImportContent(content: String?): ParsedSchedule {
        val trimmed = (content ?: "").trim()
        if (trimmed.isEmpty()) throw ParseException("导入内容为空")
        var source = trimmed
        if (source.startsWith("```")) {
            source = source
                .replace(Regex("^```(?:json|html|javascript)?\\s*", RegexOption.IGNORE_CASE), "")
                .replace(Regex("\\s*```$"), "")
        }
        val first = source.trimStart().firstOrNull()
        if (first == '{' || first == '[') {
            val value: Any? = try {
                JSONTokener(source).nextValue()
            } catch (e: Exception) {
                throw ParseException("JSON 格式错误：${e.message ?: "解析失败"}")
            }
            val obj = value as? JSONObject
            return try {
                normalizeJsonPackage(obj)
            } catch (e: ParseException) {
                if (e.message == "JSON 中没有 courses 数组" || e.message == "JSON 中没有有效课程") throw e
                throw ParseException("JSON 格式错误：${e.message}")
            }
        }
        return parseTimetable(source)
    }

    /** 供 WebView 注入脚本回传结果做二次规范化时使用。 */
    fun normalizeJsonString(json: String): ParsedSchedule = normalizeJsonPackage(JSONObject(json))
}
